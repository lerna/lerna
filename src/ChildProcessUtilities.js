"use strict";

const chalk = require("chalk");
const EventEmitter = require("events");
const execa = require("execa");
const logTransformer = require("strong-log-transformer");
const pFinally = require("p-finally");

// Keep track of how many live children we have.
let children = 0;

// This is used to alert listeners when all children have exited.
const emitter = new EventEmitter();

// when streaming children are spawned, use this color for prefix
const colorWheel = ["cyan", "magenta", "blue", "yellow", "green", "red"];
const NUM_COLORS = colorWheel.length;

function exec(command, args, opts, callback) {
  const options = Object.assign({}, opts);
  options.stdio = "pipe"; // node default

  return _spawn(command, args, options, callback);
}

function execSync(command, args, opts) {
  return execa.sync(command, args, opts).stdout;
}

function spawn(command, args, opts, callback) {
  const options = Object.assign({}, opts);
  options.stdio = "inherit";

  return _spawn(command, args, options, callback);
}

function spawnStreaming(command, args, opts, prefix, callback) {
  const options = Object.assign({}, opts);
  options.stdio = ["ignore", "pipe", "pipe"];

  const colorName = colorWheel[children % NUM_COLORS];
  const color = chalk[colorName];
  const spawned = _spawn(command, args, options, callback);

  const prefixedStdout = logTransformer({ tag: `${color.bold(prefix)}:` });
  const prefixedStderr = logTransformer({ tag: `${color(prefix)}:`, mergeMultiline: true });

  // Avoid "Possible EventEmitter memory leak detected" warning due to piped stdio
  if (children > process.stdout.listenerCount("close")) {
    process.stdout.setMaxListeners(children);
    process.stderr.setMaxListeners(children);
  }

  spawned.stdout.pipe(prefixedStdout).pipe(process.stdout);
  spawned.stderr.pipe(prefixedStderr).pipe(process.stderr);

  return spawned;
}

function getChildProcessCount() {
  return children;
}

function onAllExited(callback) {
  emitter.on("empty", callback);
}

function registerChild(child) {
  children += 1;

  pFinally(child, () => {
    children -= 1;

    if (children === 0) {
      emitter.emit("empty");
    }
  }).catch(() => {});
}

// eslint-disable-next-line no-underscore-dangle
function _spawn(command, args, opts, callback) {
  const child = execa(command, args, opts);

  registerChild(child);

  if (callback) {
    child.then(result => callback(null, result.stdout), err => callback(err));
  }

  return child;
}

exports.exec = exec;
exports.execSync = execSync;
exports.spawn = spawn;
exports.spawnStreaming = spawnStreaming;
exports.getChildProcessCount = getChildProcessCount;
exports.onAllExited = onAllExited;
