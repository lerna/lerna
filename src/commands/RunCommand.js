import NpmUtilities from "../NpmUtilities";
import Command from "../Command";
import async from "async";

export default class RunCommand extends Command {
  initialize(callback) {
    this.script = this.input[0];
    this.args = this.input.slice(1);

    if (!this.script) {
      callback(new Error("You must specify which npm script to run."));
      return;
    }

    this.packagesWithScript = this.filteredPackages
      .filter((pkg) => pkg.scripts && pkg.scripts[this.script]);

    if (!this.packagesWithScript.length) {
      callback(new Error(`No packages found with the npm script '${this.script}'`));
      return;
    }

    callback(null, true);
  }

  execute(callback) {
    this.runScriptInPackages((err) => {
      if (err) {
        callback(err);
      } else {
        this.logger.success(`Successfully ran npm script '${this.script}' in packages:`);
        this.logger.success(this.packagesWithScript.map((pkg) => `- ${pkg.name}`).join("\n"));
        callback(null, true);
      }
    });
  }

  runScriptInPackages(callback) {
    async.parallelLimit(this.packagesWithScript.map((pkg) => (cb) => {
      this.runScriptInPackage(pkg, cb);
    }), this.concurrency, callback);
  }

  runScriptInPackage(pkg, callback) {
    NpmUtilities.runScriptInDir(this.script, this.args, pkg.location, (err, stdout) => {
      this.logger.info(stdout);
      if (err) {
        this.logger.error(`Errored while running npm script '${this.script}' in '${pkg.name}'`, err);
      }
      callback(err);
    });
  }
}
