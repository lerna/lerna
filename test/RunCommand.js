import assert from "assert";
import path from "path";

import ChildProcessUtilities from "../src/ChildProcessUtilities";
import exitWithCode from "./_exitWithCode";
import initFixture from "./_initFixture";
import RunCommand from "../src/commands/RunCommand";
import stub from "./_stub";

describe("RunCommand", () => {
  let testDir;

  beforeEach(done => {
    testDir = initFixture("RunCommand/basic", done);
  });

  it("should run a command", done => {
    const runCommand = new RunCommand(["my-script"], {});

    runCommand.runValidations();
    runCommand.runPreparations();

    let calls = 0;
    stub(ChildProcessUtilities, "exec", (command, options, callback) => {
      assert.equal(command, "npm run my-script ")

      if (calls === 0) assert.deepEqual(options, { cwd: path.join(testDir, "packages/package-1"), env: process.env });
      if (calls === 1) assert.deepEqual(options, { cwd: path.join(testDir, "packages/package-3"), env: process.env });

      calls++;
      callback();
    });

    runCommand.runCommand(exitWithCode(0, done));
  });

  it("should run a command for a given scope", done => {
    const runCommand = new RunCommand(["my-script"], {scope: "package-1"});

    runCommand.runValidations();
    runCommand.runPreparations();

    const ranInPackages = [];
    stub(ChildProcessUtilities, "exec", (command, options, callback) => {
      ranInPackages.push(options.cwd.substr(path.join(testDir, "packages/").length));
      callback();
    });

    runCommand.runCommand(exitWithCode(0, () => {
      assert.deepEqual(ranInPackages, ["package-1"]);
      done();
    }));
  });
});
