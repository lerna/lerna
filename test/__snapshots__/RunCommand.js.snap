// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunCommand in a basic repo always runs env script 1`] = `
Array [
  "packages/package-1 env",
  "packages/package-4 env",
  "packages/package-2 env",
  "packages/package-3 env",
]
`;

exports[`RunCommand in a basic repo does not run a script in ignored packages 1`] = `
Array [
  "packages/package-1 my-script",
]
`;

exports[`RunCommand in a basic repo runs a script in all packages with --parallel 1`] = `
Array [
  "packages/package-1 env",
  "packages/package-2 env",
  "packages/package-3 env",
  "packages/package-4 env",
]
`;

exports[`RunCommand in a basic repo runs a script in packages 1`] = `
Array [
  "packages/package-1 my-script",
  "packages/package-3 my-script",
]
`;

exports[`RunCommand in a basic repo runs a script in packages with --stream 1`] = `
Array [
  "packages/package-1 my-script",
  "packages/package-3 my-script",
]
`;

exports[`RunCommand in a basic repo runs a script only in scoped packages 1`] = `
Array [
  "packages/package-1 my-script",
]
`;

exports[`RunCommand in a basic repo should filter packages that are not updated with --since 1`] = `
Array [
  "packages/package-3 my-script",
]
`;

exports[`RunCommand with --include-filtered-dependencies runs scoped command including filtered deps 1`] = `
Array [
  "packages/package-1 my-script",
  "packages/package-2 my-script",
]
`;
